qiskit_experiments/characterization/t1_experiment.py

```diff
@@ -0,0 +1,161 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""
+T1 Experiment class.
+"""
+
+from typing import List, Optional, Union, Tuple
+import numpy as np
+
+from qiskit.circuit import QuantumCircuit
+
+from qiskit_experiments.base_experiment import BaseExperiment
+from qiskit_experiments.base_analysis import BaseAnalysis
+from qiskit_experiments import AnalysisResult
+
+
+class T1Analysis(BaseAnalysis):
+    """T1 Experiment result analysis class."""
+
+    def _run_analysis(
+        self, experiment_data, t1_guess=None, amplitude_guess=None, offset_guess=None, **kwargs
+    ) -> Tuple[AnalysisResult, None]:
+        """
+        Calculate T1
+
+        Args:
+            experiment_data (ExperimentData): the experiment data to analyze
+            t1_guess: Optional, an initial guess of T1
+            amplitude_guess: Optional, an initial guess of the coefficient of the exponent
+            offset_guess: Optional, an initial guess of the offset
+
+        Returns:
+            The analysis result with the estimated T1
+        """
+
+        size = len(experiment_data._data)
+        delays = np.zeros(size, dtype=float)
+        means = np.zeros(size, dtype=float)
+        stddevs = np.zeros(size, dtype=float)
+
+        for i, circ in enumerate(experiment_data._data):
+            delays[i] = circ["metadata"]["delay"]
+            count0 = circ["counts"].get("0", 0)
+            count1 = circ["counts"].get("1", 0)
+            shots = count0 + count1
+            means[i] = count1 / shots
+            stddevs[i] = np.sqrt(means[i] * (1 - means[i]) / shots)
+            # problem for the fitter if one of the std points is
+            # exactly zero
+            if stddevs[i] == 0:
+                stddevs[i] = 1e-4
+
+        if t1_guess is None:
+            t1_guess = np.mean(delays)
+        if offset_guess is None:
+            offset_guess = means[-1]
+        if amplitude_guess is None:
+            amplitude_guess = means[0] - offset_guess
+
+        fit_out, fit_err, chisq = BaseAnalysis.curve_fit_wrapper(
```

Comment:
If these functions are moved to another location they can just be imported rather than accessed from base class

Reply:
done